Unable to cast object of type 'System.Compiler.TypeParameter' to type 'System.Compiler.Class'.

When I have somewhat crazy class in my project like this:
```C#
public abstract class EntityDataAccessLayerPart<TDal, TDataContext, TItemBase, TSoftDelete> : EntityDataAccessLayerPart
    where TItemBase : class
    where TSoftDelete : class, TItemBase
    where TDal : EntityDataAccessLayer<TDataContext, TItemBase, TSoftDelete>
    where TDataContext : EntityContextBase
```
When I have code contracts enabled in my project I get following error when compiling
```
1>------ Build started: Project: RFERL.EntityFramework, Configuration: Debug-Fast Any CPU ------
1>  elapsed time: 1672ms
1>C:\Program Files (x86)\Microsoft\Contracts\MsBuild\v12.0\Microsoft.CodeContracts.targets(397,5): warning : Contract reference assembly for project 'RFERL.CoreUtilities' not found. Select 'Build' or 'DoNotBuild' for Contract Reference in project settings.
1>  Reading assembly 'RFERL.EntityFramework' from 'C:\Code\common\RFERL.EntityFramework\obj\Debug-Fast\RFERL.EntityFramework.dll' resulted in errors.
1>  	Unable to cast object of type 'System.Compiler.TypeParameter' to type 'System.Compiler.Class'.
1>ccrewrite : error : Rewrite aborted due to metadata errors. Check output window
1>  elapsed time: 789,191ms
1>C:\Program Files (x86)\Microsoft\Contracts\MsBuild\v12.0\Microsoft.CodeContracts.targets(254,5): error MSB3073: The command ""C:\Program Files (x86)\Microsoft\Contracts\Bin\ccrewrite.exe" "@RFERL.EntityFrameworkccrewrite.rsp"" exited with code -1.
========== Build: 0 succeeded, 1 failed, 7 up-to-date, 0 skipped ==========

```
When I comment out part of the code like this
```C#
public abstract class EntityDataAccessLayerPart<TDal, TDataContext, TItemBase, TSoftDelete> : EntityDataAccessLayerPart
    where TItemBase : class
    where TSoftDelete : class, TItemBase
    where TDal : EntityDataAccessLayer//<TDataContext, TItemBase, TSoftDelete>
    where TDataContext : EntityContextBase
```
everything works

--- a/Foxtrot/Tests/Sources/ComplexGeneric.cs
+++ b/Foxtrot/Tests/Sources/ComplexGeneric.cs
@@ -13,8 +13,11 @@
 // THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 //
-// We test here that ?? operators which are compiled to Dup if then pop else .. by C# do not cause
-// unwarranted warnings in the extractor.
+// This is a regression test for Microsoft/CodeContracts#20:
+//
+//   Unable to cast object of type 'System.Compiler.TypeParameter' to type 'System.Compiler.Class'.
+//
+// https://github.com/Microsoft/CodeContracts/issues/20
 //
 
 using System.Diagnostics.Contracts;

--- a/Foxtrot/Tests/Sources/ComplexGeneric.cs
+++ b/Foxtrot/Tests/Sources/ComplexGeneric.cs
@@ -1,0 +1,65 @@
+// CodeContracts
+// 
+// Copyright (c) Microsoft Corporation
+// 
+// All rights reserved. 
+// 
+// MIT License
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+//
+// We test here that ?? operators which are compiled to Dup if then pop else .. by C# do not cause
+// unwarranted warnings in the extractor.
+//
+
+using System.Diagnostics.Contracts;
+
+namespace Tests.Sources
+{
+
+    // CCRewriter was unable to read this code from the IL due to an issue in CCI.
+    // This test just makes sure that the fix is in place.
+
+    public interface IIdentityObject<out TIdentity>
+    {
+    }
+
+    public abstract class DefaultDomainBLLBase<TPersistentDomainObjectBase, TDomainObjectBase, TDomainObject, TIdent>
+        where TPersistentDomainObjectBase : class, IIdentityObject<TIdent>, TDomainObjectBase
+        where TDomainObjectBase : class
+        where TDomainObject : class, TPersistentDomainObjectBase
+    {
+        protected TDomainObject GetNested<TNestedDomainObject>(TDomainObject x) where TNestedDomainObject : TDomainObject
+        {
+            return default(TDomainObject);
+        }
+    }
+
+    public abstract class DefaultSecurityDomainBLLBase<TPersistentDomainObjectBase, TDomainObjectBase, TDomainObject, TIdent>
+        : DefaultDomainBLLBase<TPersistentDomainObjectBase, TDomainObjectBase, TDomainObject, TIdent>
+        where TPersistentDomainObjectBase : class, IIdentityObject<TIdent>, TDomainObjectBase
+        where TDomainObjectBase : class
+        where TDomainObject : class, TPersistentDomainObjectBase
+    {
+    }
+
+    partial class TestMain
+    {
+        partial void Run()
+        {
+            if (!this.behave)
+            {
+                throw new System.ArgumentNullException();
+            }
+        }
+
+        public ContractFailureKind NegativeExpectedKind = ContractFailureKind.Precondition;
+        public string NegativeExpectedCondition = "Value cannot be null.";
+    }
+
+}

--- a/Foxtrot/Tests/TestInputs.xml
+++ b/Foxtrot/Tests/TestInputs.xml
@@ -119,6 +119,15 @@
      />
 
   <TestFile
+     Name="Foxtrot\Tests\Sources\ComplexGeneric.cs"
+     CompilerOptions=""
+     FoxtrotOptions=""
+     References="System.Core.dll"
+     ContractReferenceAssemblies="false"
+     BinDir="false"
+     Compiler="CS"
+     />
+  <TestFile
      Name="Foxtrot\Tests\Sources\PostAmbleDeletion.cs"
      CompilerOptions=""
      FoxtrotOptions="-assemblyMode standard"

--- a/System.Compiler/Reader.cs
+++ b/System.Compiler/Reader.cs
@@ -1798,25 +1798,34 @@ namespace System.Compiler.Metadata{
     private TypeNodeList GetTypeParametersFor(int parentIndex, Member parent){
       GenericParamRow[] genericParameters = this.tables.GenericParamTable;
       TypeNodeList types = new TypeNodeList();
-      int i = 0, n = genericParameters.Length, j = n-1;
-      bool sorted = (this.sortedTablesMask >> (int)TableIndices.GenericParam) % 2 == 1;
-      if (sorted){
-        while (i < j){
-          int k = (i+j) / 2;
-          if (genericParameters[k].Owner < parentIndex)
-            i = k+1;
-          else
-            j = k;
+      TypeNodeList savedTypes = this.currentTypeParameters;
+      this.currentTypeParameters = types;
+      try
+      {
+        int i = 0, n = genericParameters.Length, j = n-1;
+        bool sorted = (this.sortedTablesMask >> (int)TableIndices.GenericParam) % 2 == 1;
+        if (sorted){
+          while (i < j){
+            int k = (i+j) / 2;
+            if (genericParameters[k].Owner < parentIndex)
+              i = k+1;
+            else
+              j = k;
+          }
+          while (i > 0 && genericParameters[i-1].Owner == parentIndex) i--;
         }
-        while (i > 0 && genericParameters[i-1].Owner == parentIndex) i--;
+        for (int index = 0; i < n; i++, index++)
+          if (genericParameters[i].Owner == parentIndex)
+            types.Add(this.GetGenericParameter(i, index, parent));
+          else if (sorted)
+            break;
+        if (types.Count == 0) return null;
+          return types;
       }
-      for (int index = 0; i < n; i++, index++)
-        if (genericParameters[i].Owner == parentIndex)
-          types.Add(this.GetGenericParameter(i, index, parent));
-        else if (sorted)
-          break;
-      if (types.Count == 0) return null;
-      return types;
+      finally
+      {
+        this.currentTypeParameters = savedTypes;
+      }
     }
     private TypeNode GetGenericParameter(int index, int parameterListIndex, Member parent){
       GenericParamRow[] genericParameters = this.tables.GenericParamTable;

